app/waf/src/controllers/adc.controller.ts

```diff
@@ -193,4 +207,143 @@ export class AdcController {
       );
     }
   }
+
+  // TODO: schema.response not work well here.
+  // It shows the below in Example Value:
+  // {
+  //   "actionsresponse": {
+  //     "id": "11111111-2222-3333-4444-555555555555"
+  //   }
+  // }
+  @post(prefix + '/adcs/{adcId}/action', {
+    responses: {
+      '200': Schema.response(ActionsResponse, 'Adc Id for the actions.'),
+    },
+  })
+  async provision(
+    @param(Schema.pathParameter('adcId', 'ADC resource ID')) id: string,
+    @requestBody(Schema.createRequest(ActionsRequest, 'actions request'))
+    actionBody: ActionsBody,
+  ): Promise<object | undefined> {
+    let adc = await this.adcRepository.findById(id);
+
+    switch (Object.keys(actionBody)[0]) {
+      case 'create':
+        try {
+          if (adc.status !== 'NONE' && adc.status !== 'ERROR')
+            throw new HttpErrors.BadRequest(
+              'Adc status is ' +
+                adc.status +
+                ". Cannot repeat 'create' on the same ADC.",
+            );
+
+          // TODO: Create VM in async way.
+          return await this.createOn(adc).then(async () => {
+            return {id: adc.id};
+          });
+        } catch (error) {
+          throw new HttpErrors.BadRequest(error.message);
+        }
+
+      case 'delete':
+        break;
+
+      case 'setup':
+        break;
+
+      default:
+        throw new HttpErrors.BadRequest(
+          'Not supported: ' + Object.keys(actionBody)[0],
+        );
+    }
+  }
+
+  private async serialize(adc: Adc, data?: object) {
+    if (data) Object.assign(adc, data);
+    await this.adcRepository.update(adc);
+  }
+
+  private async createOn(adc: Adc): Promise<void> {
+    try {
+      await this.serialize(adc, {status: 'BUILDING'})
+        .then(async () => await this.cNet(adc))
+        .then(async () => await this.cSvr(adc));
```

Comment:
for example, if cNet() succeeds and cSvr() fails , what happens to the created net-related staff(ports)?  will they just be left un-deleted?

Reply:
To the question: yes, the created ports will be left here undeleted. 

Reason for this: 
1. The recovery logic may be complicated, so I prefer to put it separately somewhere else. 
    In fact, we already support doing that via `ready`, `name` and `status` variables. With those variables, we are able to find any allocated resources. 
2. We make it as possible that the functions(`cNet` & `cSvr`) are reentrant and stateless(save intermediate states back to DB: `this.serialize`). 
    So with the above logic, the user can recall API the second time, then we *continue* the creation without recreating already-allocated resources.

Reply:
at that moment, those ports are not bound to anything in neutron/nova, so e.g. they can be deleted in neutron (although you saved into our db). when the user calls our api for the 2nd time, they won't be created again according to "ready"/"status" (thus, no related port will be ready in neutron side). this will probably cause consistency issue between our db and neutron/nova.  seems a problem we need to handle anyway in the future.

Reply:
In the first phase, we suppose users won't manually modify the resources(allocated for & belong to F5 [well named via `portName`]) outside of waf, in which case the resources' status becomes difficult to control. 
However, we still provide the cleanup function to clean all resources. If one port is unable to cleanup, it should be considered to have been removed or re-used by the user themselves.